### PR TITLE
feat: log all matched mocks per failed testcase

### PR DESCRIPTION
**Pull request by AkashKumar7902**
## Related Issue
  - Info about Issue or bug

Closes: #1562 

#### Describe the changes you've made
if debug flag is enabled then all the mocks which are matched are logged in case of failed testcase.
Added the logic to remove unused mocks when `removeUnusedMocks` flag is enabled.

<details>
<summary>How I tested this feature ? (guide) </summary>
To test this feature I recorded two test-sets, one having 2 test-cases and other having 4 test-cases.
image

I added body.ts as a noisy field in test-set-1 which will let it to pass and test-set-0 will fail. purpose of doing this is to demonstrate that --removeUnusedMocks only works when test-set passes.

Now lets remove a test-case, say test-4.yaml. 

currently, in mocks.yaml of test-set-1, there are a total of 7 mocks

image

lets run `keploy test` command with --removeUnusedMocks flag set and see the log:
```bash
       ▓██▓▄
    ▓▓▓▓██▓█▓▄
     ████████▓▒
          ▀▓▓███▄      ▄▄   ▄               ▌
         ▄▌▌▓▓████▄    ██ ▓█▀  ▄▌▀▄  ▓▓▌▄   ▓█  ▄▌▓▓▌▄ ▌▌   ▓
       ▓█████████▌▓▓   ██▓█▄  ▓█▄▓▓ ▐█▌  ██ ▓█  █▌  ██  █▌ █▓
      ▓▓▓▓▀▀▀▀▓▓▓▓▓▓▌  ██  █▓  ▓▌▄▄ ▐█▓▄▓█▀ █▓█ ▀█▄▄█▀   █▓█
       ▓▌                           ▐█▌                   █▌
        ▓
  
version: 2-dev

🐰 Keploy: 2024-02-22T15:30:37+05:30    WARN    Delay is set to 5 seconds, incase your app takes more time to start use --delay to set custom delay
🐰 Keploy: 2024-02-22T15:30:37+05:30    INFO    Example usage: sudo -E env PATH=$PATH keploy test -c "/path/to/user/app" --delay 6
🐰 Keploy: 2024-02-22T15:30:37+05:30    INFO            {"keploy test and mock path": "/home/akash/Desktop/keploy/keploy"}
🐰 Keploy: 2024-02-22T15:30:37+05:30    INFO            {"keploy testReport path": "/home/akash/Desktop/keploy/keploy/testReports/test-run-4"}
🐰 Keploy: 2024-02-22T15:30:42+05:30    INFO    keploy initialized and probes added to the kernel.
🐰 Keploy: 2024-02-22T15:30:42+05:30    INFO    Java detected and CA already exists     {"path": "/usr/lib/jvm/java-11-openjdk-amd64/lib/security/cacerts"}
🐰 Keploy: 2024-02-22T15:30:42+05:30    INFO    Java detected and CA already exists     {"path": "/usr/lib/jvm/java-11-openjdk-amd64/lib/security/cacerts"}
🐰 Keploy: 2024-02-22T15:30:43+05:30    INFO    Keploy has hijacked the DNS resolution mechanism, your application may misbehave in keploy test mode if you have provided wrong domain name in your application code.
🐰 Keploy: 2024-02-22T15:30:43+05:30    INFO    Proxy started at port:16789
🐰 Keploy: 2024-02-22T15:30:43+05:30    INFO    starting DNS server at addr :16789
🐰 Keploy: 2024-02-22T15:30:43+05:30    INFO    running user application for    {"test-set": "test-set-0"}
PID: 175836
           [GIN-debug] [WARNING] Creating an Engine instance with the Logger and Recovery middleware already attached.

                                                                                                                      [GIN-debug] [WARNING] Running in "debug" mode. Switch to "release" mode in production.
                                                                - using env:    export GIN_MODE=release
                                                                                                        - using code:   gin.SetMode(gin.ReleaseMode)

       [GIN-debug] GET    /:param                   --> main.getURL (3 handlers)
                                                                                [GIN-debug] POST   /url                      --> main.putURL (3 handlers)
            [GIN-debug] [WARNING] You trusted all proxies, this is NOT safe. We recommend you to set a value.
                                                                                                             Please check https://pkg.go.dev/github.com/gin-gonic/gin#readme-don-t-trust-all-proxies for details.
                                                                    [GIN-debug] Listening and serving HTTP on :8080
                                                                                                                   🐰 Keploy: 2024-02-22T15:30:43+05:30      INFO            {"no of test cases": 2, "test-set": "test-set-0"}
🐰 Keploy: 2024-02-22T15:30:48+05:30    INFO    starting test for of    {"test case": "test-1", "test set": "test-set-0"}
[GIN] 2024/02/22 - 15:30:48 | 200 |    2.255878ms |             ::1 | POST     "/url"
Testrun failed for testcase with id: "test-1"

--------------------------------------------------------------------

+-------------------------------------------------------------------------------------------------------------+
|                                                DIFFS TEST-1                                                 |
+-------------------------------------------------------------------------------------------------------------+
|                     EXPECT HEADER                    |                   ACTUAL HEADER                      |
| -----------------------------------------------------+----------------------------------------------------- |
|                                                      |                                                      |
|                                                                                                             |
|                      EXPECT BODY                     |                    ACTUAL BODY                       |
| -----------------------------------------------------+----------------------------------------------------- |
|    {                                                 |  {                                                   |
|   -  "ts": 1.7085915317954831e+18,                   | +  "ts": 1.7085960481977405e+18,                     |
|      "url": "http://localhost:8080/Lhr4BWAi"         |    "url": "http://localhost:8080/Lhr4BWAi"           |
|    }                                                 |  }                                                   |
|                                                      |                                                      |
|                                                                                                             |
+-------------------------------------------------------------------------------------------------------------+
🐰 Keploy: 2024-02-22T15:30:48+05:30    INFO            {"matched mocks": ["mock-3"]}
🐰 Keploy: 2024-02-22T15:30:48+05:30    INFO    result  {"testcase id": "test-1", "testset id": "test-set-0", "passed": "false"}
🐰 Keploy: 2024-02-22T15:30:48+05:30    INFO    starting test for of    {"test case": "test-2", "test set": "test-set-0"}
[GIN] 2024/02/22 - 15:30:48 | 303 |     558.955µs |             ::1 | GET      "/Lhr4BWAi"
Testrun passed for testcase with id: "test-2"

--------------------------------------------------------------------

🐰 Keploy: 2024-02-22T15:30:48+05:30    INFO    result  {"testcase id": "test-2", "testset id": "test-set-0", "passed": "true"}
🐰 Keploy: 2024-02-22T15:30:48+05:30    INFO    test report for test-set-0:     {"name: ": "report-1", "path: ": "/home/akash/Desktop/keploy/keploy/report-1"}

 <=========================================> 
  TESTRUN SUMMARY. For testrun with id: "test-set-0"
        Total tests: 2
        Total test passed: 1
        Total test failed: 1
 <=========================================> 

🐰 Keploy: 2024-02-22T15:30:48+05:30    INFO    keploy has initiated the shutdown of the user application.
🐰 Keploy: 2024-02-22T15:30:48+05:30    INFO    keploy terminated user application
🐰 Keploy: 2024-02-22T15:30:48+05:30    INFO    running user application for    {"test-set": "test-set-1"}
PID: 175921
           [GIN-debug] [WARNING] Creating an Engine instance with the Logger and Recovery middleware already attached.

                                                                                                                      [GIN-debug] [WARNING] Running in "debug" mode. Switch to "release" mode in production.
                                                                - using env:    export GIN_MODE=release
                                                                                                        - using code:   gin.SetMode(gin.ReleaseMode)

       [GIN-debug] GET    /:param                   --> main.getURL (3 handlers)
                                                                                [GIN-debug] POST   /url                      --> main.putURL (3 handlers)
            [GIN-debug] [WARNING] You trusted all proxies, this is NOT safe. We recommend you to set a value.
                                                                                                             Please check https://pkg.go.dev/github.com/gin-gonic/gin#readme-don-t-trust-all-proxies for details.
                                                                    [GIN-debug] Listening and serving HTTP on :8080
                                                                                                                   🐰 Keploy: 2024-02-22T15:30:48+05:30      INFO            {"no of test cases": 3, "test-set": "test-set-1"}
🐰 Keploy: 2024-02-22T15:30:53+05:30    INFO    starting test for of    {"test case": "test-1", "test set": "test-set-1"}
[GIN] 2024/02/22 - 15:30:53 | 200 |     1.27849ms |             ::1 | POST     "/url"
Testrun passed for testcase with id: "test-1"

--------------------------------------------------------------------

🐰 Keploy: 2024-02-22T15:30:53+05:30    INFO    result  {"testcase id": "test-1", "testset id": "test-set-1", "passed": "true"}
🐰 Keploy: 2024-02-22T15:30:53+05:30    INFO    starting test for of    {"test case": "test-2", "test set": "test-set-1"}
[GIN] 2024/02/22 - 15:30:53 | 303 |     724.771µs |             ::1 | GET      "/Lhr4BWAi"
Testrun passed for testcase with id: "test-2"

--------------------------------------------------------------------

🐰 Keploy: 2024-02-22T15:30:53+05:30    INFO    result  {"testcase id": "test-2", "testset id": "test-set-1", "passed": "true"}
🐰 Keploy: 2024-02-22T15:30:53+05:30    INFO    starting test for of    {"test case": "test-3", "test set": "test-set-1"}
[GIN] 2024/02/22 - 15:30:53 | 200 |     455.969µs |             ::1 | POST     "/url"
Testrun passed for testcase with id: "test-3"

--------------------------------------------------------------------

🐰 Keploy: 2024-02-22T15:30:53+05:30    INFO    result  {"testcase id": "test-3", "testset id": "test-set-1", "passed": "true"}
🐰 Keploy: 2024-02-22T15:30:53+05:30    INFO    removed unused mocks from mock file     {"test-set": "test-set-1"}
🐰 Keploy: 2024-02-22T15:30:53+05:30    INFO    test report for test-set-1:     {"name: ": "report-2", "path: ": "/home/akash/Desktop/keploy/keploy/report-2"}

 <=========================================> 
  TESTRUN SUMMARY. For testrun with id: "test-set-1"
        Total tests: 3
        Total test passed: 3
        Total test failed: 0
 <=========================================> 

🐰 Keploy: 2024-02-22T15:30:53+05:30    INFO    keploy has initiated the shutdown of the user application.

 <=========================================> 
  COMPLETE TESTRUN SUMMARY. 
        Total tests: 5
        Total test passed: 4
        Total test failed: 1

        Test Suite Name         Total Test      Passed          Failed

        "test-set-0"            2               1               1
        "test-set-1"            3               3               0
<=========================================> 

🐰 Keploy: 2024-02-22T15:30:53+05:30    INFO    test run completed      {"passed overall": false}
🐰 Keploy: 2024-02-22T15:30:53+05:30    INFO    Exiting keploy program gracefully.
🐰 Keploy: 2024-02-22T15:30:53+05:30    INFO    keploy terminated user application
🐰 Keploy: 2024-02-22T15:30:53+05:30    INFO    eBPF resources released successfully...
🐰 Keploy: 2024-02-22T15:30:53+05:30    INFO    Dns server stopped
🐰 Keploy: 2024-02-22T15:30:53+05:30    INFO    proxy stopped...
```

Looking at the logs you can notice that unused mocks are removed only after test-set-1, and not after failed test-set-0 as discussed above. 
`
🐰 Keploy: 2024-02-22T15:30:53+05:30    INFO    removed unused mocks from mock file     {"test-set": "test-set-1"}
`

also now mocks.yaml has only 5 mocks instead of 6.
image

</details>

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Please let us know if any test cases are added

Please describe the tests(if any). Provide instructions how its affecting the coverage.

#### Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)
A clear and concise description of it.

## Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.

## Screenshots (if any)

|        Original         |          Updated           |
|:-----------------------:|:--------------------------:|
| **original screenshot** | <b>updated screenshot </b> |

### CODE REVIEW

File: pkg/hooks/loader.go [hunk 1]
```diff
@@ -182,6 +183,7 @@ func (h *Hook) SetConfigMocks(m []*models.Mock) {
 
 func (h *Hook) UpdateConfigMock(oldMock *models.Mock, newMock *models.Mock) bool {
 	isUpdated := h.configMocks.update(oldMock.TestModeInfo, newMock.TestModeInfo, newMock)
+	h.matchedMocks = append(h.matchedMocks, newMock.Name)
```
Comment (charankamarapu) [inline pkg/hooks/loader.go:186]: shouldn't these be inside a mutex because there are many go routine which are deleting and updating mocks..?

File: pkg/service/test/test.go [hunk 2]
```diff
@@ -509,10 +509,12 @@ func (t *tester) SimulateRequest(cfg *SimulateRequestConfig) {
 		testPass, testResult := t.testHttp(*cfg.Tc, resp, cfg.NoiseConfig, cfg.IgnoreOrdering)
 
 		if !testPass {
+			t.logger.Debug("", zap.Any("matched mocks", cfg.LoadedHooks.GetMatchedMocks()))
```
Comment (charankamarapu) [inline pkg/service/test/test.go:512]: Please add the logic for deleting unused mocks and a flag to enable this feature.

**Review by charankamarapu — COMMENTED** [review]
Please address the comments and fill the PR description

File: pkg/service/test/test.go (hunk 2 above)
Comment (AkashKumar7902) [reply pkg/service/test/test.go:512]: will the unused mocks be removed at end of test-set run or test case run ? also we need to rewrite the mock file right ?

Comment (AkashKumar7902) [reply pkg/service/test/test.go:515]: done!

File: pkg/service/test/test.go [hunk 3]
```diff
@@ -509,10 +512,12 @@ func (t *tester) SimulateRequest(cfg *SimulateRequestConfig) {
...
```
Comment (shivamsouravjha) [inline pkg/service/test/test.go:515]: incase if debug flag is enabled, the logs are just too much and we will loose even the test run summary and other context.

It's better to use info log as enabling debug flag 
image

Comment (AkashKumar7902) [reply pkg/service/test/test.go]: done!

**Comment by shivamsouravjha** [comment]
Did you check if any new race condition is introduced ?

File: pkg/hooks/loader.go [hunk 4]
```diff
@@ -225,14 +244,34 @@ func (h *Hook) GetConfigMocks() ([]*models.Mock, error) {
...
+}
+
+func (h *Hook) GetUsedMocks() ([]*models.Mock, error) {
+	var matchedMocks []*models.Mock
+	for matchedMock := range h.persistentMatchedMocks {
+		matchedMocks = append(matchedMocks, matchedMock)
+	}
+	return matchedMocks, nil
+}
+
```
Comment (shivamsouravjha) [inline pkg/hooks/loader.go:270]: why maintain a map when an array can be maintained and array is used in the end?

File: pkg/platform/interface.go [hunk 5]
```diff
@@ -2,12 +2,16 @@ package platform
...
 	"context"
+
+	"go.keploy.io/server/pkg/models"
 )
 
 type TestCaseDB interface {
 	WriteTestcase(tc KindSpecifier, ctx context.Context, filters KindSpecifier) error
 	WriteMock(tc KindSpecifier, ctx context.Context) error
 
+	RewriteMocks(mocks []*models.Mock, testSet string) error
```
Comment (shivamsouravjha) [inline pkg/platform/interface.go:13]: can UpdateMock be a better alternative?

File: pkg/hooks/loader.go (hunk 4 above)
Comment (AkashKumar7902) [reply pkg/hooks/loader.go:270]: This map serves the purpose of set data structure, in case where addition of same entries happens an array would store copies but persistentMatchedMocks would store only one

File: pkg/platform/yaml/yaml.go [hunk 6]
```diff
@@ -358,14 +362,33 @@ func (ys *Yaml) WriteMock(mockRead platform.KindSpecifier, ctx context.Context)
...
+		if err != nil {
+			return err
+		}
+		if i == 0 {
+			err = ys.Write(ys.MockPath, "mocks", mockYaml, true)
+		} else {
+			err = ys.Write(ys.MockPath, "mocks", mockYaml, false)
+		}
+		if err != nil {
+			return err
```
Comment (shivamsouravjha) [inline pkg/platform/yaml/yaml.go:386]: instead of this can the whole file be deleted before hand ? and so that write function isn't refactored and this is a very specific case of just checking for a single element of the array.

File: pkg/platform/interface.go (hunk 5 above)
Comment (AkashKumar7902) [reply pkg/platform/interface.go]: Done

File: pkg/hooks/loader.go [hunk 7]
```diff
@@ -225,14 +244,34 @@ func (h *Hook) GetConfigMocks() ([]*models.Mock, error) {
...
+	h.persistentMatchedMocks[mock] = true
 	return isDeleted
 }
 
 func (h *Hook) DeleteConfigMock(mock *models.Mock) bool {
 	isDeleted := h.configMocks.delete(mock.TestModeInfo)
+	h.matchedMocks = append(h.matchedMocks, mock.Name)
+	h.persistentMatchedMocks[mock] = true
 	return isDeleted
 }
```
Comment (shivamsouravjha) [inline pkg/hooks/loader.go:258]: incase if the mock isn't deleted (meaning isn't matched) then also its gets pushed in the map for delete, and update.

We should only put the mock that's matched right?

File: pkg/platform/yaml/yaml.go (hunk 6 above)
Comment (AkashKumar7902) [reply pkg/platform/yaml/yaml.go]: shall I change update it to use the delete file logic ?

Comment (shivamsouravjha) [reply pkg/platform/yaml/yaml.go]: Would be a better option, we already have all the mocks we need we can simply delete it because in truncate that's the operation we're doing.

File: pkg/hooks/loader.go [hunk 8]
```diff
@@ -162,6 +165,18 @@ func (h *Hook) AppendMocks(m *models.Mock, ctx context.Context) error {
 	return nil
 }
 
+func (h *Hook) RemoveUnusedMocks(testSet string) error {
+	mocks, err := h.GetUsedMocks()
```
Comment (shivamsouravjha) [inline pkg/hooks/loader.go:169]: also is it safe to directly get usedmocks without using mutex? like can there be any race condition?

Comment (AkashKumar7902) [reply pkg/hooks/loader.go:169]: This function is only called when a test set run is completed, so at a time only control will access this function.

File: pkg/hooks/loader.go (hunk 7 above)
Comment (AkashKumar7902) [reply pkg/hooks/loader.go:258]: delete functions are used when a match takes place 
https://github.com/keploy/keploy/blob/e117ea1ff0113a408fb73a0c27a9e8404b0d825a/pkg/proxy/integrations/grpcparser/request_to_response_mapping.go#L76
https://github.com/keploy/keploy/blob/e117ea1ff0113a408fb73a0c27a9e8404b0d825a/pkg/proxy/integrations/httpparser/match.go#L75
https://github.com/keploy/keploy/blob/e117ea1ff0113a408fb73a0c27a9e8404b0d825a/pkg/proxy/integrations/mongoparser/match.go#L55

File: pkg/platform/yaml/yaml.go (hunk 6 above)
Comment (AkashKumar7902) [reply pkg/platform/yaml/yaml.go]: done!

File: pkg/hooks/loader.go (hunk 7 above)
Comment (shivamsouravjha) [reply pkg/hooks/loader.go:258]:
```func (db *treeDb) delete(key interface{}) bool {
	db.mutex.Lock()
	defer db.mutex.Unlock()
	_, found := db.rbt.Get(key)
	if !found {
		return false
	}
	db.rbt.Remove(key)
	return true
}
```

```isDeleted := h.DeleteTcsMock(mock)
		if !isDeleted {
			continue
		}
```


If some other test case matched to the mock , it will be deleted from memory , isdeleted for this mock would be false and we will continue looking for next most suitable mock. Incase we go with overwriting without checking is deleted, we will be triggering map related functions multiple time

File: pkg/platform/yaml/yaml.go [hunk 9]
```diff
@@ -366,6 +366,25 @@
 	return nil
 }
 
+func (ys *Yaml) UpdateMocks(mocks []*models.Mock, testSet string) error {
+	ys.MockPath = filepath.Join(ys.MockPath, testSet)
+	err := os.Remove(filepath.Join(ys.MockPath, "mocks.yaml"))
```
[the remark this answers is not in the document]
Comment (shivamsouravjha) [reply pkg/platform/yaml/yaml.go:371]: use util.validate path as this causes security vulnerability

File: pkg/platform/yaml/yaml.go [hunk 10]
```diff
@@ -366,6 +366,25 @@ func (ys *Yaml) WriteMock(mockRead platform.KindSpecifier, ctx context.Context)
...
+	if err != nil {
+		return nil
+	}
```
Comment (shivamsouravjha) [inline pkg/platform/yaml/yaml.go:374]: return nil incase of error?

File: pkg/platform/yaml/yaml.go [hunk 11]
```diff
@@ -103,8 +103,8 @@ func (ys *Yaml) Write(path, fileName string, docRead platform.KindSpecifier) err
 	if err != nil {
 		return err
 	}
-
-	file, err := os.OpenFile(yamlPath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, os.ModePerm)
+	var file *os.File
+	file, err = os.OpenFile(yamlPath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, os.ModePerm)
```
Comment (shivamsouravjha) [inline pkg/platform/yaml/yaml.go:107]: can be reverted to original I guess

**Review by shivamsouravjha — CHANGES_REQUESTED** [review]

File: pkg/hooks/loader.go (hunk 7 above)
Comment (shivamsouravjha) [reply pkg/hooks/loader.go:258]: any update on this?

Comment (AkashKumar7902) [reply pkg/hooks/loader.go:258]: > Incase we go with overwriting without checking is deleted, we will be triggering map related functions multiple time

couldn't understand this

Comment (shivamsouravjha) [reply pkg/hooks/loader.go:258]: mock match doesn't guarantee mock being deleted. If some other call got matched with the best matched mock it will be deleted (eg: we're looking for a call-1 , mock5 is best match but it's already used by call-5 , so it will be removed from the tree, in that can call-1 will look for another mock.

Incase if we ignore the isdeleted var we would be doing map[mock5] = true twice. And this map operation takes more time complexity, so better to trigger map[mock5] = true only once.

Comment (AkashKumar7902) [reply pkg/hooks/loader.go]: i have made the changes. please have a look

**Comment by AkashKumar7902** [comment]
regarding the race condition, there were total 4 found data races, but none of them point to the files changed in this PR 
https://pastebin.com/JncRn2k6

File: pkg/hooks/loader.go [hunk 12]
```diff
@@ -174,14 +189,20 @@ func (h *Hook) SetTcsMocks(m []*models.Mock) {
 func (h *Hook) SetConfigMocks(m []*models.Mock) {
 	h.configMocks.deleteAll()
 	for index, mock := range m {
+		h.persistentMatchedMocks[mock.Name] = mock
 		mock.TestModeInfo.SortOrder = index
```
Comment (shivamsouravjha) [inline pkg/hooks/loader.go:193]: why is this needed?

Comment (AkashKumar7902) [reply pkg/hooks/loader.go:193]: Without is `persistentMatchedMocks` will contain only the tcsMocks

File: pkg/hooks/loader.go [hunk 13]
```diff
@@ -174,14 +190,21 @@ func (h *Hook) SetTcsMocks(m []*models.Mock) {
 func (h *Hook) SetConfigMocks(m []*models.Mock) {
 	h.configMocks.deleteAll()
 	for index, mock := range m {
+		h.persistentMatchedMocks[mock.Name] = false
 		mock.TestModeInfo.SortOrder = index
 		mock.TestModeInfo.Id = index
 		h.configMocks.insert(mock.TestModeInfo, mock)
 	}
```
Comment (shivamsouravjha) [inline pkg/hooks/loader.go:197]: why setting all to true ?

File: pkg/hooks/loader.go [hunk 14]
```diff
@@ -225,14 +248,67 @@ func (h *Hook) GetConfigMocks() ([]*models.Mock, error) {
...
 	return isDeleted
 }
 
 func (h *Hook) DeleteConfigMock(mock *models.Mock) bool {
 	isDeleted := h.configMocks.delete(mock.TestModeInfo)
+	if isDeleted {
+		h.persistentMatchedMocks[mock.Name] = false
+	}
 	return isDeleted
 }
```
Comment (shivamsouravjha) [inline pkg/hooks/loader.go:263]: incase if a mock is matched you're setting it to false , why?

File: pkg/hooks/loader.go (hunk 13 above)
Comment (AkashKumar7902) [reply pkg/hooks/loader.go:197]: I have added a comment describing the purpose of the key & value of `persistentMatchedMocks` 

https://github.com/keploy/keploy/blob/3faff242965634d853967ce4aed8182f9e33ad79/pkg/hooks/loader.go#L55


previously in test.go of service layer, i was fetching the matched mocks from array and resetting it. now, for persistentMatchedMocks to serve the purpose of removed `matchedMocks` array, I have implemented `getMatchedMocks` function. this function will look through the entries of persistentMatchedMocks and return the mocks which have value set to true (meaning it is not a config mock and it is not used till yet)
https://github.com/keploy/keploy/blob/3faff242965634d853967ce4aed8182f9e33ad79/pkg/service/test/util.go#L468

File: pkg/hooks/loader.go [hunk 15]
```diff
@@ -174,14 +190,21 @@ func (h *Hook) SetTcsMocks(m []*models.Mock) {
...
 
 func (h *Hook) UpdateConfigMock(oldMock *models.Mock, newMock *models.Mock) bool {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
 	isUpdated := h.configMocks.update(oldMock.TestModeInfo, newMock.TestModeInfo, newMock)
+	if isUpdated {
+		h.DeletePersistentMock(oldMock.Name)
+	}
+	h.persistentMatchedMocks[newMock.Name] = false
```
Comment (shivamsouravjha) [inline pkg/hooks/loader.go:207]: so if mock is not  updated we would set the new mock as well in flags?

File: pkg/hooks/loader.go (hunk 14 above)
Comment (AkashKumar7902) [reply pkg/hooks/loader.go:263]: https://github.com/keploy/keploy/pull/1573#discussion_r1497570055

File: pkg/service/test/util.go [hunk 16]
```diff
@@ -463,6 +465,16 @@ func FilterMocks(tc *models.TestCase, m []*models.Mock, logger *zap.Logger) ([]*
 	return filteredMocks, unFilteredMocks
 }
 
+func GetMatchedMocks(persistentMocks map[string]bool) []string {
+	var matchedMocks []string
+	for k, v := range persistentMocks {
+		if v {
+			matchedMocks = append(matchedMocks, k)
+		}
+	}
```
Comment (shivamsouravjha) [inline pkg/service/test/util.go:474]: better names can be used

File: pkg/hooks/loader.go [hunk 17]
```diff
@@ -225,14 +248,67 @@ func (h *Hook) GetConfigMocks() ([]*models.Mock, error) {
...
+	return h.persistentMatchedMocks
+}
+ 
+func (h *Hook) DeletePersistentMock(mockName string) bool {
+	if _, ok := h.persistentMatchedMocks[mockName]; ok {
+		delete(h.persistentMatchedMocks, mockName)
+		return true
+	}
+	return false
+}
```
Comment (shivamsouravjha) [inline pkg/hooks/loader.go:310]: probably a mutex on this?

**Review by shivamsouravjha — CHANGES_REQUESTED** [review]

File: pkg/service/test/util.go (hunk 16 above)
Comment (AkashKumar7902) [reply pkg/service/test/util.go]: Done

File: pkg/hooks/loader.go (hunk 17 above)
Comment (AkashKumar7902) [reply pkg/hooks/loader.go]: Done

File: pkg/service/test/util.go [hunk 18]
```diff
@@ -463,6 +465,16 @@ func FilterMocks(tc *models.TestCase, m []*models.Mock, logger *zap.Logger) ([]*
 	return filteredMocks, unFilteredMocks
 }
 
+func GetMatchedMocks(persistentMocks map[string]bool) []string {
+	var matchedMocks []string
+	for mockName, isTcsAndNotUsed := range persistentMocks {
+		if isTcsAndNotUsed {
```
Comment (shivamsouravjha) [inline pkg/service/test/util.go:471]: isTcsUnused

Comment (AkashKumar7902) [reply pkg/service/test/util.go]: done!

File: pkg/hooks/loader.go [hunk 19]
```diff
@@ -174,6 +190,7 @@ func (h *Hook) SetTcsMocks(m []*models.Mock) {
 func (h *Hook) SetConfigMocks(m []*models.Mock) {
 	h.configMocks.deleteAll()
 	for index, mock := range m {
+		h.persistentMatchedMocks[mock.Name] = false
 		mock.TestModeInfo.SortOrder = index
```
Comment (shivamsouravjha) [inline pkg/hooks/loader.go:194]: for sake of consistency UpdatePersistentMatchedMock

Comment (AkashKumar7902) [reply pkg/hooks/loader.go:194]: done!

File: pkg/hooks/loader.go [hunk 20]
```diff
@@ -225,14 +246,75 @@ func (h *Hook) GetConfigMocks() ([]*models.Mock, error) {
...
+		}
+		ConfigMocks = append(ConfigMocks, configMock)
+	}
+	mocks := append(ConfigMocks, TcsMocks...)
+	for _, mock := range mocks {
+		if _, ok := h.persistentMatchedMocks[mock.Name]; ok {
+			matchedMocks = append(matchedMocks, mock)
+		}
+	}
+	return matchedMocks, nil
```
Comment (shivamsouravjha) [inline pkg/hooks/loader.go:295]: there are config calls that go in between like health check or etc for that sake its better to sort all mocks on name itself

**Review by shivamsouravjha — CHANGES_REQUESTED** [review]

File: pkg/hooks/loader.go [hunk 21]
```diff
@@ -225,14 +247,82 @@ func (h *Hook) GetConfigMocks() ([]*models.Mock, error) {
...
+		}
+	}
+	SortMocksByName(matchedMocks)
+	return matchedMocks, nil
+}
+
+func SortMocksByName(mocks []*models.Mock) {
+	sort.SliceStable(mocks, func(i, j int) bool {
+		return mocks[i].Name < mocks[j].Name
+	})
```
[the remark this answers is not in the document]
Comment (AkashKumar7902) [reply pkg/hooks/loader.go:314]: I have made the change. please have a look

**Review by shivamsouravjha — APPROVED** [review]
LGTM!

File: pkg/hooks/loader.go [hunk 22]
```diff
@@ -175,6 +194,7 @@ func (h *Hook) SetTcsMocks(m []*models.Mock) {
 func (h *Hook) SetConfigMocks(m []*models.Mock) {
 	h.configMocks.deleteAll()
 	for index, mock := range m {
+		h.UpdatePersistentMatchedMock(mock.Name, false)
 		mock.TestModeInfo.SortOrder = index
```
Comment (shivamsouravjha) [inline pkg/hooks/loader.go:198]: since we're keep only the mocks that are used better to remove this?

**Review by shivamsouravjha — CHANGES_REQUESTED** [review]

File: pkg/hooks/loader.go (hunk 22 above)
Comment (AkashKumar7902) [reply pkg/hooks/loader.go:198]: doing this would mean no config mocks in persistentMatchedMocks and only tcsMocks will get written in mocks.yaml
